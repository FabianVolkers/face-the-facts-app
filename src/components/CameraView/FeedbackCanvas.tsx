import React from 'react';
import AbstractCanvas from './AbstractCanvas';

import {NormalizedFace} from '@tensorflow-models/blazeface';

class FeedbackCanvas extends AbstractCanvas {
    draw = (predictions: NormalizedFace[]): void => {
		const ua = navigator.userAgent.toLowerCase();
		const isSafari = ua.includes('safari') && ua.indexOf('chrome') === -1;
		const ctx = this.ref.current?.getContext('2d') as CanvasRenderingContext2D;
		if (ctx) {
			ctx.clearRect(0, 0, ctx.canvas.width, ctx.canvas.height);

			// Render a rectangle over each detected face.
			ctx.beginPath();
			ctx.strokeStyle = 'white';
			ctx.lineWidth = 6;

			// draw full screen clockwise, then face bbox counter clockwise
			// to darken everything but the face
			if (!isSafari) {
				ctx.rect(0, 0, ctx.canvas.width, ctx.canvas.height);
			}

			predictions.forEach((prediction) => {
				const start: [number, number] = prediction.topLeft as [number, number];
				const end: [number, number] = prediction.bottomRight as [number, number];

				/* const probability = prediction.probability as number;
				const prob = (probability * 100).toPrecision(5).toString(); */

				const size = [end[0] - start[0], end[1] - start[1]];

				if (!isSafari) {
					// counter clockwise
					ctx.rect(end[0], start[1], -size[0], size[1]);
				} else {
					ctx.rect(start[0], start[1], size[0], size[1]);
				}
			});
			ctx.stroke();
			if (!isSafari && predictions.length > 0) {
				ctx.fillStyle = 'rgba(0,0,0,0.5)';
				ctx.fill();
			}
		}
	};
    
}

export default FeedbackCanvas;