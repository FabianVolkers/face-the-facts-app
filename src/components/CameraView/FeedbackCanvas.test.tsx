import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import 'jest-canvas-mock';
import FeedbackCanvas from './FeedbackCanvas';

let container: HTMLDivElement | null = null;

beforeEach(() => {
	// setup a DOM element as a render target
	container = document.createElement('div');
	document.body.appendChild(container);
});

afterEach(() => {
	// cleanup on exiting
	if (container !== null) {
		unmountComponentAtNode(container);
		container.remove();
		container = null;
	}
});
describe('integration test', () => {
	it('renders with correct size', () => {
		// given
		const ref: React.RefObject<FeedbackCanvas> = React.createRef();
		const width: number = 100;
		const height: number = 100;

		//when
		act(() => {
			render(<FeedbackCanvas ref={ref}></FeedbackCanvas>, container);
			ref.current?.init(width, height);
		});

		// then
		if (container !== null) {
			const canvas = container.getElementsByTagName('canvas')[0];
			expect(canvas.width).toBe(width);
			expect(canvas.height).toBe(height);
		}
	});
	it('draws without errors', () => {
		// given
		const ref: React.RefObject<FeedbackCanvas> = React.createRef();
		const predictions: Array<{ topLeft: [number, number]; bottomRight: [number, number] }> = [
			{ topLeft: [20, 50], bottomRight: [80, 100] },
		];
		const width: number = 100;
		const height: number = 100;

		// when
		act(() => {
			render(<FeedbackCanvas ref={ref}></FeedbackCanvas>, container);
			ref.current?.init(width, height);
		});

		// then
		expect(() => ref.current?.draw(predictions)).not.toThrow();
	});

	it('draws detections correctly', () => {
		// given
		const ref: React.RefObject<FeedbackCanvas> = React.createRef();
		const predictions: Array<{ topLeft: [number, number]; bottomRight: [number, number] }> = [
			{ topLeft: [20, 50], bottomRight: [80, 100] },
		];
		const width: number = 100;
		const height: number = 100;

		// when
		act(() => {
			render(<FeedbackCanvas ref={ref}></FeedbackCanvas>, container);
			ref.current?.init(width, height);
			ref.current?.draw(predictions);
		});

		// then
		const path = ref.current?.ref.current?.getContext('2d')?.__getPath;
		expect(path).toMatchSnapshot();
	});
});
