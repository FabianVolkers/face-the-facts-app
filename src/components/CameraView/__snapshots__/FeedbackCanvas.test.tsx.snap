// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration test draws detections correctly 1`] = `[Function]`;
